plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/jvmmodel/DispatchMethodCompileStrategy.java

```diff
@@ -205,11 +209,14 @@ protected void generateActualDispatchCall(JvmOperation dispatchOperation, JvmOpe
 		a.append(")");
 	}
 
-	/**
-	 * TODO: we assume that names from the inferred JVM model are the same in the Java code. Does that assumption
-	 * hold?
-	 */
 	protected String getVarName(JvmIdentifiableElement ex, IAppendable appendable) {
-		return ex.getSimpleName();
+		final String name = ex.getSimpleName();
+		if (name == null) {
+			return "__unknown__";
+		} else if (keywords.isJavaKeyword(name)) {
+			return name+"_";
```

Comment:
I've just tried it.
Input:
class Foo {
	def dispatch foo(Object assert, Object assert_) {}
        def dispatch foo(String assert, Object assert_) {}
}

Output:
public Object foo(final Object assert_, final Object assert__1) {
    if (assert_ instanceof String) {
      return _foo((String)assert_, **assert_**);
    } else if (assert_ != null) {
      return _foo(assert_, **assert_**);
    } else {
      throw new IllegalArgumentException("Unhandled parameter types: " +
        Arrays.<Object>asList(assert_, **assert_**).toString());
    }
  }

It seems that bogus second argument is used inside the dispatch method.